ark/schema/structure/optional.ts

```diff
@@ -95,23 +95,36 @@ export const Optional = {
 	Node: OptionalNode
 }
 
+const isPrimitive = (value: unknown): value is Primitive =>
+	typeof value === "object" ? value === null : typeof value !== "function"
+
 export const assertDefaultValueAssignability = (
 	node: BaseRoot,
 	value: unknown,
 	key = ""
 ): unknown => {
-	const out = node.in(value)
+	if (!isPrimitive(value) && typeof value !== "function")
+		throwParseError(writeNonPrimitiveNonFunctionDefaultValueMessage(key))
+
+	const out = node.in(typeof value === "function" ? value() : value)
 	if (out instanceof ArkErrors)
 		throwParseError(writeUnassignableDefaultValueMessage(out.message, key))
+
 	return value
 }
 
 export const writeUnassignableDefaultValueMessage = (
 	message: string,
 	key = ""
-): string => `Default value${key && ` for key ${key}`} ${message}`
+): string =>
+	`Default value${key && ` for key ${key}`} is not assignable: ${message}`
```

Comment:
I feel like this would create redundant messages in a lot of cases?

Did we try just:


```ts
`Default ${key && ` for key ${key}`} ${message}`
```

Reply:
Oh I wasn't thinking of errors at a path like this. Keep it the way it was then I might tweak it later.